File: NeetCodeIO/235LowestCommonAncestorOfABinarySearchTree/LowestCommonAncestorOfABinarySearchTree.cpp
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        // DFS Space optimized -> Time:O(n), Space:O(1)
        TreeNode* curr = root;
        while (curr) {
            if (p->val < curr->val && q->val < curr->val) {
                curr = curr->left;
            }
            else if (p->val > curr->val && q->val > curr->val) {
                curr = curr->right;
            }
            else {
                return curr;
            }
        }
        return curr;

        // // DFS -> Time:O(n), Space:O(h)
        // if (p->val < root->val && q->val < root->val) {
        //     return lowestCommonAncestor(root->left, p, q);
        // }
        // else if (p->val > root->val && q->val > root->val) {
        //     return lowestCommonAncestor(root->right, p, q);
        // }
        // return root;
    }
};

File: NeetCodeIO/235LowestCommonAncestorOfABinarySearchTree/LowestCommonAncestorOfABinarySearchTree.py
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # DFS Space optimized -> Time:O(n), Space:O(1)
        curr = root
        while curr:
            if p.val < curr.val and q.val < curr.val:
                curr = curr.left
            elif p.val > curr.val and q.val > curr.val:
                curr = curr.right
            else:
                return curr
        
        # # DFS -> Time:O(n), Space:O(h)
        # if p.val > root.val and q.val > root.val:
        #     return self.lowestCommonAncestor(root.right, p, q)
        # elif p.val < root.val and q.val < root.val:
        #     return self.lowestCommonAncestor(root.left, p, q)
        # return root
